core/src/main/kotlin/org/evomaster/core/search/gene/MapGene.kt

```diff
@@ -91,21 +91,46 @@ class MapGene<T>(
         return maxSize > 0
     }
 
-    override fun standardMutation(randomness: Randomness, apc: AdaptiveParameterControl, allGenes: List<Gene>) {
+    override fun candidatesInternalGenes(randomness: Randomness, apc: AdaptiveParameterControl, allGenes: List<Gene>, selectionStrategy: SubsetGeneSelectionStrategy, enableAdaptiveGeneMutation: Boolean, additionalGeneMutationInfo: AdditionalGeneSelectionInfo?): List<Gene> {
+        if(!isMutable()){
+            throw IllegalStateException("Cannot mutate a immutable array")
+        }
+        if ( elements.isEmpty() || elements.size > maxSize){
+            return listOf()
+        }
+        val p = when(selectionStrategy){
+            SubsetGeneSelectionStrategy.ADAPTIVE_WEIGHT -> {
+                if(additionalGeneMutationInfo?.impact != null
+                        && additionalGeneMutationInfo.impact is MapGeneImpact
+                        && additionalGeneMutationInfo.impact.sizeImpact.noImprovement.any { it.value < 2 } //if there is recent improvement by manipulating size
+                ){
+                    0.3 // increase probability to mutate size
+                }else MODIFY_SIZE
+            }
+            else ->{
+                MODIFY_SIZE
+
+            }
+        }
```

Comment:
you have used exactly the same code before. would be good to somehow have it in its own function (utility class, or interface with defaults, or modify the gene hierarchy) to avoid copy&paste of same code